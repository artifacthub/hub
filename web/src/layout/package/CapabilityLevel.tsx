import classnames from 'classnames';
import { isNull, isUndefined } from 'lodash';
import React from 'react';
import { FaCheck } from 'react-icons/fa';
import { MdInfo } from 'react-icons/md';

import Modal from '../common/Modal';
import SmallTitle from '../common/SmallTitle';
import styles from './CapabilityLevel.module.css';

interface Props {
  capabilityLevel?: string;
}

const LEVELS = ['Basic Install', 'Seamless Upgrades', 'Full Lifecycle', 'Deep Insights', 'Auto Pilot'];

const CapabilityLevel = (props: Props) => {
  if (isUndefined(props.capabilityLevel) || isNull(props.capabilityLevel)) return null;

  const activeLevel = LEVELS.findIndex((level: string) => level === props.capabilityLevel);

  return (
    <div>
      <div className="d-flex flex-row align-items-center">
        <SmallTitle text="Capability Level" />
        <Modal
          buttonType="btn-link btn-sm px-0 pb-0 text-secondary"
          buttonContent={<MdInfo />}
          className="d-none d-lg-inline-block mt-1 ml-2"
          header={<div className="h5 m-0">Capability level</div>}
        >
          <div className="my-3 mw-100">
            <img
              src="/static/media/capability-level-diagram.svg"
              alt="Capability Level Diagram"
              className="capability-level-diagram mw-100"
            />
          </div>
        </Modal>
      </div>

      <div className="mb-3 position-relative">
        {LEVELS.map((level: string, index: number) => (
          <div
            key={`capabilityLevel-${index}`}
            className={`d-flex flex-row align-items-center my-2 position-relative stepWrapper ${styles.stepWrapper}`}
          >
            <div
              data-testid="capabilityLevelStep"
              className={classnames('rounded-circle text-center mr-2 textLight step', styles.step, {
                [`activeStep ${styles.activeStep}`]: activeLevel >= index,
              })}
            >
              {activeLevel >= index && <FaCheck />}
            </div>
            <small
              className={classnames({
                'text-muted': activeLevel < index,
              })}
            >
              {level}
            </small>
          </div>
        ))}
      </div>
    </div>
  );
};

export default CapabilityLevel;
