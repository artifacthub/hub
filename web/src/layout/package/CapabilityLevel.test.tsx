import { fireEvent, render } from '@testing-library/react';
import React from 'react';

import CapabilityLevel from './CapabilityLevel';

describe('DefaultDetails', () => {
  it('creates snapshot', () => {
    const result = render(<CapabilityLevel capabilityLevel="Deep Insights" />);
    expect(result.asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders component', () => {
      const { getByText, getAllByTestId } = render(<CapabilityLevel capabilityLevel="Deep Insights" />);

      expect(getByText('Capability Level')).toBeInTheDocument();

      expect(getByText('Basic Install')).toBeInTheDocument();
      expect(getByText('Seamless Upgrades')).toBeInTheDocument();
      expect(getByText('Full Lifecycle')).toBeInTheDocument();
      expect(getByText('Deep Insights')).toBeInTheDocument();
      expect(getByText('Auto Pilot')).toBeInTheDocument();

      const steps = getAllByTestId('capabilityLevelStep');
      expect(steps[0]).toHaveClass('activeStep');
      expect(steps[1]).toHaveClass('activeStep');
      expect(steps[2]).toHaveClass('activeStep');
      expect(steps[3]).toHaveClass('activeStep');
      expect(steps[4]).not.toHaveClass('activeStep');
    });

    it('opens modal', () => {
      const { getByTestId, getByRole, getByAltText } = render(<CapabilityLevel capabilityLevel="Deep Insights" />);

      const modal = getByRole('dialog');
      expect(modal).not.toHaveClass('d-block');

      const btn = getByTestId('openModalBtn');
      fireEvent.click(btn);

      expect(modal).toHaveClass('d-block');
      expect(getByAltText('Capability Level Diagram')).toBeInTheDocument();
    });
  });
});
