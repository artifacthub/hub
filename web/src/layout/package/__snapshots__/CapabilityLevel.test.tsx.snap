// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultDetails creates snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="d-flex flex-row align-items-center"
    >
      <div
        class="mt-2 mb-1"
        data-testid="smallTitle"
      >
        <small
          class="card-title text-muted text-uppercase"
        >
          <span>
            Capability Level
          </span>
        </small>
      </div>
      <div
        class="d-none d-lg-inline-block mt-1 ml-2"
      >
        <div
          class="position-relative buttonWrapper"
        >
          <button
            class="font-weight-bold text-uppercase position-relative btn btn-block btn btn-link btn-sm px-0 pb-0 text-secondary"
            data-testid="openModalBtn"
            type="button"
          >
            <div
              class="d-flex align-items-center justify-content-center"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
                />
              </svg>
            </div>
          </button>
        </div>
        <div
          class="modal modal"
          role="dialog"
        >
          <div
            class="modal-dialog modal-lg modal-dialog-centered modal-dialog-scrollable"
          >
            <div
              class="modal-content content undefined"
            >
              <div
                class="modal-header d-flex flex-row align-items-center header"
              >
                <div
                  class="h5 m-0"
                >
                  Capability level
                </div>
                <button
                  class="close"
                  data-testid="closeModalBtn"
                  type="button"
                >
                  <span
                    aria-hidden="true"
                  >
                    Ã—
                  </span>
                </button>
              </div>
              <div
                class="modal-body p-4 h-100 d-flex flex-column"
              />
              <div
                class="modal-footer p-3"
              >
                <button
                  class="btn btn-secondary"
                  data-testid="closeModalFooterBtn"
                  type="button"
                >
                  Close
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="mb-3 position-relative"
    >
      <div
        class="d-flex flex-row align-items-center my-2 position-relative stepWrapper stepWrapper"
      >
        <div
          class="rounded-circle text-center mr-2 textLight step step activeStep activeStep"
          data-testid="capabilityLevelStep"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
            />
          </svg>
        </div>
        <small
          class=""
        >
          Basic Install
        </small>
      </div>
      <div
        class="d-flex flex-row align-items-center my-2 position-relative stepWrapper stepWrapper"
      >
        <div
          class="rounded-circle text-center mr-2 textLight step step activeStep activeStep"
          data-testid="capabilityLevelStep"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
            />
          </svg>
        </div>
        <small
          class=""
        >
          Seamless Upgrades
        </small>
      </div>
      <div
        class="d-flex flex-row align-items-center my-2 position-relative stepWrapper stepWrapper"
      >
        <div
          class="rounded-circle text-center mr-2 textLight step step activeStep activeStep"
          data-testid="capabilityLevelStep"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
            />
          </svg>
        </div>
        <small
          class=""
        >
          Full Lifecycle
        </small>
      </div>
      <div
        class="d-flex flex-row align-items-center my-2 position-relative stepWrapper stepWrapper"
      >
        <div
          class="rounded-circle text-center mr-2 textLight step step activeStep activeStep"
          data-testid="capabilityLevelStep"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
            />
          </svg>
        </div>
        <small
          class=""
        >
          Deep Insights
        </small>
      </div>
      <div
        class="d-flex flex-row align-items-center my-2 position-relative stepWrapper stepWrapper"
      >
        <div
          class="rounded-circle text-center mr-2 textLight step step"
          data-testid="capabilityLevelStep"
        />
        <small
          class="text-muted"
        >
          Auto Pilot
        </small>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
